src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,101 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double PERCENTAGE_GROWTH = 1.5D;
+    private int size;
+    private Object[] elements;
+
+    public ArrayList() {
+        elements = new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        checkArraySize();
+        elements[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        isExistOrNext(index);
+        checkArraySize();
+        System.arraycopy(elements, index, elements, index + 1, size - index);
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        isExist(index);
+        return (T) elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        isExist(index);
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        isExist(index);
+        T removedElement = (T) elements[index];
+        System.arraycopy(elements, index + 1, elements, index, size - index);
+        size--;
+        return removedElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (elements[i] == element || elements[i] != null && elements[i].equals(element)) {
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException("Element missing: " + element.toString());
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void isExistOrNext(int index) {
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Element missing for given index "
+                    + index);
+        }
+    }
+
+    private void isExist(int index) {
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Element missing for given index "
+                    + index);
+        }
+    }
```

Comment:
Consider using single method for checking index
```suggestion
    private void isExistOrNext(int index, int bound) {
        if (index > bound || index < 0) { // you can use size or size-1 as bound
            throw new ArrayListIndexOutOfBoundsException("Element missing for given index "
                    + index);
        }
    }
```

Reply:
I believe it's better to keep both methods, as the first one is used when adding a new element and checks indices within the range of length+1, whereas in the second case, we're simply checking if an element with the desired index exists. If we were to use the same method for both cases, conflicts would arise, for instance, when we try to retrieve an element at the index of length+1.